src/workerd/server/server.c++

```diff
@@ -1365,25 +1372,152 @@ public:
 
     kj::Own<WorkerInterface> getActor(kj::String id,
         IoChannelFactory::SubrequestMetadata metadata) {
+      kj::String idCopy = kj::str(id);
       return newPromisedWorkerInterface(service.waitUntilTasks,
-          getActorThenStartRequest(kj::mv(id), kj::mv(metadata)));
+          getActorThenStartRequest(kj::mv(id), kj::mv(metadata))).attach(
+              kj::defer([this, idCopy=kj::mv(idCopy)]() {
+            // Once this WorkerInterface is dropped, we need to run some cleanup tasks.
+            KJ_IF_SOME(entry, actors.find(idCopy)) {
+              entry->clientDropped();
```

Comment:
I've been looking into this, it feels like we get stronger guarantees but I'm running into some trouble. We' want to track the `Own<WorkerInterface>` returned by `getActor`, but that method is actually returning a `PromisedWorkerInterface`.

Reply:
I've been looking into this, it feels like we would get stronger guarantees but I'm running into some trouble. 

We want to track the `Own<WorkerInterface>` returned by `getActor`, but `getActor` is actually returning a `PromisedWorkerInterface`. By the time we construct and return the object, we don't necessarily have an `ActorContainerRef`  to `attach()` to it (ex. on first request). We might not even have an entry in `actors` (i.e. an `ActorContainer`).

We know we have an `ActorContainerRef` once `getActorThenStartRequest` resolves, so ideally we could setup another promise where we: 
- fork the `getActorThenStartRequest` promise and wait for it, 
- get a strong ref to `ActorContainerRef`, then
- attach it to the `Own<WorkerInterface>` that's returned by `getActor`. 

However, I'm not able to `addBranch()` that promise since `WorkerInterface` doesn't implement `addRef()` (in this case, the `WorkerInterface` is a wrapped `WorkerEntrypoint`, so not Refcounted).

I've also tried attaching it to `service.startRequest()` when we're returning from `getActorThenStartRequest`, since that's the real `Own<WorkerInterface>` and it's moved to the `worker` property in the `PromisedWorkerInterface` (in the ctor). Unfortunately it's dropped very early on; I suspect during the `kj::mv` or after we exit the promise continuation scope.

---

Will keep looking but am curious if I'm overlooking something obvious here?

Reply:
I think what you'd do is have `getActorImpl()` return a pair of an `Actor` and an `ActorContainerRef`. `getActorThenStartRequest()` would then attach that ref to the `WorkerInterface` it returns. `getActorImpl()` would construct the ref if it doesn't already exist, immediately after constructing the actor if it doesn't already exist.

Reply:
> getActorThenStartRequest() would then attach that ref to the WorkerInterface it returns

I gave that a try on Friday, but rather than returning a pair of `Actor` and `ActorContainerRef` I just stored the `kj::Own<ActorContainerRef>` in `ActorContainer`, then moved + attached to the `WorkerInterface` returned by `getActorThenStartRequest()`. Will give it another go in the morning but I'm worried that `Own<WorkerInterface>` is dropped too early (possibly during a move in the `PromisedWorkerInterface` constructor).

Reply:
Unstuck now, had misinterpreted what was going on. Was assuming that if we already have a `Worker::Actor` we must also have an `ActorContainerRef`. Obviously if the client is dropped we could still have a `Worker::Actor` and no `ActorContainerRef`...

Reply:
> Also, since we can now evict on a per actor basis, should we just erase exactly at the 70 sec mark, or retain the mass eviction that can take anywhere from 70-140 sec?

We should match what happens in prod. We could consider changing both, but not sure if there's any particular reason to do so.